### ISSUE TITLE
[Docs]: Precache json files and handle dynamic Pages with Nextjs

### ISSUE THREAD

**Issue by guilatrova** [serwist/serwist#215]
### Provide environment information

I created a dedicated repository just for this: https://github.com/guilatrova/pwa-next-dynamic-routes
Available for demo here: https://testpwanextjsofflinejsons.vercel.app/

**Dependencies:** 

```json
{
    "@serwist/next": "9.0.9",
    "next": "15.0.1",
    "react": "19.0.0-rc-69d4b800-20241021",
    "react-dom": "19.0.0-rc-69d4b800-20241021"
  }
```

I have the following pages:

- `app/page.tsx` (index dir)
- `app/books/[book]/[bookId]/page.tsx` (dynamic route that reads data from `public`)

I have the following public JSON content hosted under `public/json/data/books`:

- `b1/1.json`
- `b1/2.json`
- `b1/3.json`
- `b2/1.json`
- `b2/2.json`

### Describe the problem

After a few tests with Serwist I learned that it precaches all the json data by default (which is fine for this example).

But it doesn't happen for the `app/books/[book]/[bookId]/page.tsx` routes, and also **I don't think it would be a good idea to precache each chapter page individually**.

### Expected behavior

Since I have the json precached I wish I could somehow precache a "default offline chapter page" for it and then dynamically load each json (that is precached) straight from the cache.

I'm uncertain on how to handle "online dynamic routes" like https://testpwanextjsofflinejsons.vercel.app/books/b1/1 to render to some sort of: `/offline/book/chapter/b1/1`.

In summary:

| Aspect | Current Behavior | Expected Behavior |
|---|---|---|
| Dynamic Chapter Routes | Don't precache individually | Meets expectation ✅  |
| JSON files | Precache all of them | Meets expectation ✅  |
| Serving Chapter Routes Offline | Can't handle even having the json precached | Should use json to fill page content ❌   |

My question is... How to achieve that? I looked around the docs but it's unclear whether I want a strategy, a plugin or something else.

### Links
Repository: https://github.com/guilatrova/pwa-next-dynamic-routes
Hosted Demo: https://testpwanextjsofflinejsons.vercel.app/
Single Book Chapter Page: https://testpwanextjsofflinejsons.vercel.app/books/b1/1

**Comment by DuCanhGH** [comment]
Before answering your question, I'd like to say that `[book]/[bookId]` seems like... a bad naming convention. Perhaps you were looking for `[book]/[chapter]` instead? Took me a good minute to understand what you were saying :)

> But it doesn't happen for the app/books/[book]/[bookId]/page.tsx routes, and also I don't think it would be a good idea to precache each chapter page individually.

Yeah, having a service worker with a SSR site can be a little bit difficult. Usually I'd encourage using search params, but that doesn't seem to play out well with SEO.

I think you can look into caching the pages with Serwist as the user hovers over links to the chapters. Just a suggestion though! For dynamic pages the likes of this, there aren't many ways besides precaching all the combinations, adding an offline fallback, or having a bit worse UX (users have to visit pages before they are available offline).

To add an offline fallback for the pages, do this:

```ts
const urlsToCache = ["/books/offline"] as const;

self.addEventListener("install", (event) => {
  event.waitUntil(
    Promise.all(
      urlsToCache.map((entry) => {
        const request = serwist.handleRequest({
          request: new Request(entry),
          event,
        });
        return request;
      }),
    ),
  );
});

serwist.setCatchHandler(async ({ event, request, url }) => {
  if (request.destination === "document" && url.origin === location.origin && url.pathname.startsWith("/books")) {
    const match = serwist.handleRequest({ event, request: new Request("/books/offline") });
    return match ? await match : Response.error();
  }
  return Response.error();
});
```

You can also use the `Serwist` class's `fallbacks` option, but that precaches the "/offline" page, which is always synchronized with the service worker. Not sure if you'd prefer that.

In the "/offline" page, use search params to generate the desired HTML.

**Comment by guilatrova** [comment]
Thank you very much.

I did what you suggested and it worked gracefully to get the offline page working seamlessly!

If anyone else is interested in looking into a sample, here's the commit: https://github.com/guilatrova/pwa-next-dynamic-routes/commit/1e0092d1ac09f70e3c98eeba08f8ae869f768502

@DuCanhGH Can you also explain how I can disable the automatic pre-caching of JSON files? There're some directories I want to skip but I see serwist decides to pull everything no matter what (?)

I'm planning on handling it on demand.

**Comment by guilatrova** [comment]
I figured out how to cache json only on demand :)

https://github.com/guilatrova/pwa-next-dynamic-routes/commit/77175eb7d194bb90763ab2b4367688777ec9c6e1

**Comment by DuCanhGH** [comment]
@guilatrova :D glad you figured it out.

I'd recommend opting out with `["**/*", "!**/*.json"]` instead though.

**Comment by guilatrova** [comment]
> I'd recommend opting out with ["**/*", "!**/*.json"] instead though.

It doesn't work... It keeps pulling the json data.
Not a problem for me but I felt it would be useful to share in case anyone attempts something similar.

**Comment by DuCanhGH** [comment]
@guilatrova ...ah. We use Isaacs's `node-glob`, so it doesn't quite work like that. Gonna have to add some new options soon, then.

**Comment by electromidz** [comment]
@DuCanhGH I have similar problem just on dynamic route I have for example [phone] route and in this specific dynamic route I got error and if I change this to phone all things work correctly, have any suggest on dynamic routing?
also I use like this and also I saw the @guilatrova  project but for my case did not work correctly
**sw.ts**
`import { Serwist } from 'serwist';

import { CUSTOM_CACHE } from './caching';

import type { PrecacheEntry, SerwistGlobalConfig } from 'serwist';

// This declares the value of `injectionPoint` to TypeScript.
// `injectionPoint` is the string that will be replaced by the
// actual precache manifest. By default, this string is set to
// `"self.__SW_MANIFEST"`.
declare global {
  interface WorkerGlobalScope extends SerwistGlobalConfig {
    __SW_MANIFEST: (PrecacheEntry | string)[] | undefined;
  }
}

declare const self: ServiceWorkerGlobalScope;

const serwist = new Serwist({
  precacheEntries: self.__SW_MANIFEST,
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  runtimeCaching: CUSTOM_CACHE,
});

const DEFAULT_URLS = ['/', '/auth'];

self.addEventListener('install', (event) => {
  event.waitUntil(
    Promise.all(
      DEFAULT_URLS.map((entry) => {
        const request = serwist.handleRequest({
          request: new Request(entry),
          event,
        });
        return request;
      })
    )
  );
});

serwist.setCatchHandler(async ({ event, request, url }) => {
  if (request.destination === 'document' && url.origin === location.origin && url.pathname.startsWith('/auth')) {
    const match = serwist.handleRequest({ event, request: new Request('/auth') });
    return match ? await match : Response.error();
  }
  return Response.error();
});

serwist.addEventListeners();`

I pass one matcher like this:
`      {
        matcher: ({ sameOrigin, url: { pathname } }) => {
          // Exclude /api/auth/callback/* to fix OAuth workflow in Safari without having
          // an impact on other environments
          // The above route is the default for next-auth, you may need to change it if
          // your OAuth workflow has a different callback route.
          // Issue: https://github.com/shadowwalker/next-pwa/issues/131#issuecomment-821894809
          // TODO(ducanhgh): Investigate Auth.js's "/api/auth/*" failing when we allow them
          // to be cached (the current behaviour).
          if (!sameOrigin || pathname.startsWith('/app/(auth)/auth/*')) {
            return false;
          }

          if (pathname.startsWith('/auth/')) {
            return true;
          }

          return false;
        },
        method: 'GET',
        handler: new NetworkFirst({
          cacheName: 'auth',
          plugins: [
            new ExpirationPlugin({
              maxEntries: 16,
              maxAgeSeconds: 24 * 60 * 60, // 24 hours
              maxAgeFrom: 'last-used',
            }),
          ],
          networkTimeoutSeconds: 10, // fallback to cache if API does not response within 10 seconds
        }),
      },`
      
![Screenshot from 2024-11-09 18-49-06](https://github.com/user-attachments/assets/04e5b87d-8644-4966-8c92-f7722ff0f3a2)

and I got this error 
![image](https://github.com/user-attachments/assets/042c10c1-0c4c-4556-a128-1afd4d512261)
![Screenshot from 2024-11-09 18-50-07](https://github.com/user-attachments/assets/77f9a0e0-de37-4da5-b2d3-ef479ba21772)

**Comment by DuCanhGH** [comment]
@electromidz seems that it's failing to fetch those JS files. Are you sure they are actually available on the server?

**Comment by electromidz** [comment]
@DuCanhGH If I remove dynamic and use phone instead of [phone] all things work correctly If you need some sandbox to just tell me.
